### PR TITLE
#168781681 -User can reset password

### PR DESCRIPTION
**Pull request by sabin18**
#### What does this PR do?
- user can Reset Password via email.

#### Description of Task to be completed?
- add send email verification function using SendGrid
- add reset password function 

#### How should this be manually tested?
- clone the repository : git clone https://github.com/andela/caret-bn-backend.git
- checkout to branch ft-reset-password-168781681
- do npm install
- add SENDGRID_API_KEY in your .env
- Go to ```/api/v1/users/forgotpassword``` to send Email  for link to reset password
- after go to the link you got in your email :`` https://caret-bn-backend.herokuapp.com/api/v1/users/resetpassword`` to reset password  or `` /api/v1//users/resetpassword`` on your localhost .
- copy the link from your browser paste it in your Postman and use the PATCH method .
-  forgot password endpoint ``:POST: /api/v1/users/forgotpassword `` :to get link for reset password
-  reset password endpoint : ``:PATCH: /api/v1/users/resetpassword ``:to reset password 

#### Any background context you want to provide?
`N/A`

#### What are the relevant pivotal tracker stories?
[#168781681 ](https://www.pivotaltracker.com/n/projects/2400651)

#### Screenshots (if appropriate)
Screen Shot 2019-10-10 at 21 26 53
Screen Shot 2019-10-11 at 15 14 08
Screen Shot 2019-10-16 at 09 43 37

#### Questions:
`N/A`

### CODE REVIEW

**Review by alainmateso — CHANGES_REQUESTED** [review]
Good job sending a reset email but when I click on the reset password button, it returns this message:

"status": 404,
"message": "Sorry this route does not exist !"
therefore I cannot fully test the feature . Kindly consider making sure that the functionality fully works!

**Review by nignanthomas — CHANGES_REQUESTED** [review]
Hey @sabin18, great job so far, keep it coming
However, the implementation of the feature is not complete and not secured.

You need to send a token created with the **generateToken** utils to create a token based on the user found by email together with the email. Then the user copies the token and add a token field to the **/resetpassword** request body that you decode in the patch to make it match the user email and decoded user from the token.
Only then, you can change the password.

If that does not make sense reach out tome

File: src/routes/api/index.js [hunk 1]
```diff
@@ -5,6 +5,8 @@ import swaggerRoute from '../swagger-doc';
 
 const router = new Router();
 router.use('/users', usersRoutes);
+// router.use('/resetpassword/', usersRoutes);
+// router.use('/forgotpassword', usersRoutes);
```
Comment (nignanthomas) [inline src/routes/api/index.js]: Remove the commented lines

File: src/routes/api/users.js [hunk 2]
```diff
@@ -114,8 +116,83 @@ const router = express.Router();
...
+ *   forgotpassword:
+ *     type: object
+ *     properties:
+ *       email:
+ *         type: string
+ *         example: krinkun09@gmail.com
+ */
+/**
+ * @swagger
+ * /forgotpassword:
```
Comment (nignanthomas) [inline src/routes/api/users.js]: swagger route should be **users/forgotpassword**

File: src/routes/api/users.js [hunk 3]
```diff
@@ -114,8 +116,83 @@ const router = express.Router();
...
+ *         type: string
+ *         example: krinkun09@gmail.com
+ *       password:
+ *         type: string
+ *         format: string
+ *         example: Pa55W0rd
+ */
+/**
+ * @swagger
+ * /resetpassword:
```
Comment (nignanthomas) [inline src/routes/api/users.js]: swagger route should be **users/resetpassword/{token}** and have a parameter **"in:path"** for users to be able to pass in the **token**
Look at Eric's swagger doc for the verify token

File: src/controllers/userController.js [hunk 4]
```diff
@@ -68,6 +71,62 @@ export default class UserController {
...
+      const verifyToken = EmailToken.ResetToken(user);
+      sgMail.setApiKey(process.env.SENDGRID_API_KEY);
+      const msg = {
+        to: email,
+        from: 'BarefootNomad@gmail.com',
+        subject: 'Barefoot Nomad Resetpassword Link',
+        html: `<div style="font-family:Avenir,Helvetica,sans-serif;box-sizing:border-box;padding:35px;">
+        <h1 style="color: #848484;">Barefoot Nomad</h1>
+        <p style="font-family:Avenir,Helvetica,sans-serif;box-sizing:border-box;color:#74787e;font-size:16px;line-height:1.5em;margin-top:0;text-align:left">Welcome Back,<br> We are happy to be with you. Please you can reset your password now.<br> Click the button below to reset your password.</p>
+        <p><a style="background-color: #3097d1; border: 2px solid #3097d1; padding: 8px; color: #fff; font-size: 16px; text-decoration: none;cursor: pointer;" href="${req.protocol}://${req.headers.host}/api/v1/users/resetpassword/${verifyToken}">Reset passowrd</a>
```
Comment (nignanthomas) [inline src/controllers/userController.js:93]: I am still not convinced by you sending a link that tells me **404, route not found**. As a user what do I do when I get this? I think it is broken. Reach out to @zellymiriam  as soon as possible @sabin18, if she agrees with that implementation, but I am not convinced

**Review by nignanthomas — CHANGES_REQUESTED** [review]

**Review by alainmateso — APPROVED** [review]

**Review by nignanthomas — APPROVED** [review]

**Review by ngireric123 — APPROVED** [review]

File: src/utils/EmailToken.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    {
+      expiresIn: '1h',
+    }
+  );
+  return token;
+};
+// Access token required for a user
+const UseraccessRequired = (req, res, next) => {
+  const { token } = req.params;
+  const decodedToken = jwt.decode(token, process.env.JWT_SECRET);
```
Comment (nakiwuge) [inline src/utils/EmailToken.js:35]: Add a new line

File: src/utils/EmailToken.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        return responseUtil(res, 400, strings.users.error.EXPERED);
+      }
+      if (!user || decodedToken.payload.send !== true) {
+        return responseUtil(res, 403, strings.users.error.ANAUTHORIZED);
+      }
+    });
+  next();
+};
+
+
```
Comment (nakiwuge) [inline src/utils/EmailToken.js]: Remove this line

File: src/validation/validateResetpassword.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import Joi from '@hapi/joi';
+
+const validateResetpassword = data => {
+  const schema = Joi.object({
+    email: Joi.string().email({ minDomainSegments: 2 }).message('email field should be a valid email address. e.g: johndoe@gmail.com.').required(),
+    newpassword: Joi.string().regex(/^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#$%^&*])(?=.{8,})/).message('password field should contain at least 8 characters, at least 1 lowercase, 1 uppercase and 1 number and a special character.').required(),
+  });
+  return schema.validate(data, { abortEarly: false });
+};
```
Comment (nakiwuge) [inline src/validation/validateResetpassword.js:10]: Add a new line

File: src/utils/EmailToken.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+import jwt from 'jsonwebtoken';
+import moment from 'moment';
+import dotenv from 'dotenv';
+import models from '../database/models';
+import responseUtil from './responseUtil';
+import strings from './stringsUtil';
+
+dotenv.config();
+
+const ResetToken = user => {
```
Comment (nakiwuge) [inline src/utils/EmailToken.js:12]: I believe there is already a function for generating a token. Why don`t you use the same function?

**Review by nakiwuge — CHANGES_REQUESTED** [review]
@sabin18 good job. I realized you left out the `confirmPassword` field. Please add it. Also I have left some comments you should check out.

**Review by nakiwuge — APPROVED** [review]

**Pull request merged**